File: magus_core/c++-test/equiv_test.cc

/// @file equiv_test.cc
/// @brief equiv_test の実装ファイル
/// @author Yusuke Matsunaga (松永 裕介)
///
/// Copyright (C) 2018 Yusuke Matsunaga
/// All rights reserved.


#include "gtest/gtest.h"
#include "EquivMgr.h"
#include "ym/BnNetwork.h"
#include "ym/Range.h"


BEGIN_NAMESPACE_MAGUS

TEST(EquivTest, EquivTest)
{
  BnNetwork network1;
  BnNetwork network2;

  string filename1 = "C499.blif";
  string path1 = DATAPATH + filename1;
  bool stat1 = read_blif(network1, path1);
  ASSERT_TRUE( stat1 );

  int ni = network1.input_num();
  int no = network1.output_num();
  vector<int> input1_list(ni);
  vector<int> output1_list(no);
  for ( int i: Range(ni) ) {
    input1_list[i] = network1.input_id(i);
  }
  for ( int i: Range(no) ) {
    output1_list[i] = network1.output_id(i);
  }

  string filename2 = "C1355.blif";
  string path2 = DATAPATH + filename2;
  bool stat2 = read_blif(network2, path2);
  ASSERT_TRUE( stat2 );
  ASSERT_TRUE( network2.input_num() == ni );
  ASSERT_TRUE( network2.output_num() == no );

  vector<int> input2_list(ni);
  vector<int> output2_list(no);
  for ( int i: Range(ni) ) {
    input2_list[i] = network2.input_id(i);
  }
  for ( int i: Range(no) ) {
    output2_list[i] = network2.output_id(i);
  }

  EquivMgr eqmgr(1000);
  vector<SatBool3> eq_stats;
  SatBool3 ans = eqmgr.check(network1, input1_list, output1_list,
			     network2, input2_list, output2_list,
			     eq_stats);
  EXPECT_EQ( SatBool3::True, ans );
  for ( int i: Range(no) ) {
    EXPECT_EQ( SatBool3::True, eq_stats[i] );
  }
}

END_NAMESPACE_MAGUS

File: magus_core/py-test/equiv_test.py
#! /usr/bin/env python3

### @file equiv_test.py
### @brief equiv のテスト
### @author Yusuke Matsunaga (松永 裕介)
###
### Copyright (C) 2018 Yusuke Matsunaga
### All rights reserved.

import unittest
import os
from magus_core import BnNetwork
from magus_core import EquivMgr
from magus_core import SatBool3

# EquivMgr のテスト用クラス
class EquivTest(unittest.TestCase) :

    def runTest(self) :
        self.test1()
        self.test2()

    def test1(self) :
        TESTDATA_DIR = os.environ.get('TESTDATA_DIR')

        filename1 = os.path.join(TESTDATA_DIR, 'C499.blif')
        network1 = BnNetwork.read_blif(filename1)
        filename2 = os.path.join(TESTDATA_DIR, 'C1355.blif')
        network2 = BnNetwork.read_blif(filename2)

        eqmgr = EquivMgr(1)
        stat, stat_list = eqmgr.check_by_order(network1, network2)
        self.assertEqual( stat, SatBool3._True )
        for stat1 in stat_list :
            self.assertEqual( stat1, SatBool3._True )

    def test2(self) :
        TESTDATA_DIR = os.environ.get('TESTDATA_DIR')

        filename1 = os.path.join(TESTDATA_DIR, 'C499.blif')
        network1 = BnNetwork.read_blif(filename1)
        filename2 = os.path.join(TESTDATA_DIR, 'C1355.blif')
        network2 = BnNetwork.read_blif(filename2)

        ni = network1.input_num
        no = network1.output_num
        input2_list = [ i for i in range(ni) ]
        output2_list = [ i for i in range(no) ]

        eqmgr = EquivMgr(1)
        stat, stat_list = eqmgr.check(network1, network2, input2_list, output2_list)
        self.assertEqual( stat, SatBool3._True )
        for stat1 in stat_list :
            self.assertEqual( stat1, SatBool3._True )

    def test3(self) :
        TESTDATA_DIR = os.environ.get('TESTDATA_DIR')

        filename1 = os.path.join(TESTDATA_DIR, 'C499.blif')
        network1 = BnNetwork.read_blif(filename1)
        filename2 = os.path.join(TESTDATA_DIR, 'C499_reordered.blif')
        network2 = BnNetwork.read_blif(filename2)

        ni = network1.input_num
        no = network1.output_num

        input_map = dict()
        for pos in range(ni) :
            id1 = network1.input_id(pos)
            node = network1.node(id1)
            input_map[node.name] = pos

        output_map = dict()
        for pos in range(no) :
            id1 = network1.output_id(pos)
            node = network2.node(id1)
            output_map[node.name] = pos

        input2_list = []
        for pos in range(ni) :
            id2 = network2.input_id(pos)
            node2 = network2.node(id2)
            self.assertTrue( node2.name in input_map )
            pos1 = input_map[node2.name]
            input2_list.append(pos1)

        output2_list = []
        for pos in range(no) :
            id2 = network2.output_id(pos)
            node2 = network2.node(id2)
            self.assertTrue( node2.name in output_map )
            pos1 = output_map[node2.name]
            output2_list.append(pos1)

        eqmgr = EquivMgr(1)
        stat, stat_list = eqmgr.check(network1, network2, input2_list, output2_list)
        self.assertEqual( stat, SatBool3._True )
        for stat1 in stat_list :
            self.assertEqual( stat1, SatBool3._True )

if __name__ == '__main__' :

    unittest.main()
